contracts/PeeranhaToken.sol

```diff
@@ -53,54 +92,105 @@ contract PeeranhaToken is ERC20Upgradeable, ERC20PausableUpgradeable, ERC20Cappe
-    require(RewardLib.getPeriod(CommonLib.getTimestamp()) > period, "This period isn't ended yet!");
+    require(msg.sender == user || msg.sender == owner(), "get_reward_security");  // unitTest
+    require(RewardLib.getPeriod(CommonLib.getTimestamp()) > period + 1, "period_not_ended");
 
     require(
       !statusRewardContainer.statusReward[user][period].isPaid,
-      "You already picked up this reward."
+      "reward_was_pick_up."
     );
 
     statusRewardContainer.statusReward[user][period].isPaid = true;
     RewardLib.WeekReward memory weekReward = peeranha.getWeekRewardContainer(period);
     
     uint256 poolToken = getPool(weekReward, period);
-    if (poolTokens[period] != poolToken) poolTokens[period] = poolToken;
+    if (poolTokens[period] == 0 && poolToken != poolTokens[period]) {
+      poolTokens[period] = poolToken;
+    }
     uint256 userReward = getUserReward(weekReward, user, period, poolToken);
 
-    require(userReward != 0, "No reward for you in this period");
+    require(userReward != 0, "no_reward");
 
     emit GetReward(user, period);
-    _mint(user, userReward * FRACTION);
+    _mint(user, userReward);
+  }
+
+  function setStake(address user, uint256 stakeTokens) external {
+    require(stakeTokens <= balanceOf(user), "wrong_stack");
+    uint16 nextPeriod = RewardLib.getPeriod(CommonLib.getTimestamp()) + 1;
+    TokenLib.StakeUserContainer storage stakeUserContainer = userPeriodStake.userPeriodStake[user];
+
+    TokenLib.StakeTotal storage stakeTotal = stakeTotalContainer.stakeTotals[nextPeriod];
+    if (stakeTotalContainer.stakeChangePeriods.length != 0 && stakeTotalContainer.stakeChangePeriods[stakeTotalContainer.stakeChangePeriods.length - 1] != nextPeriod) {
+      uint16 previousPeriod = stakeTotalContainer.stakeChangePeriods[stakeTotalContainer.stakeChangePeriods.length - 1];
+      stakeTotal.totalStakedAmount = stakeTotalContainer.stakeTotals[previousPeriod].totalStakedAmount;
+      stakeTotal.stakingUsersCount = stakeTotalContainer.stakeTotals[previousPeriod].stakingUsersCount;
+      stakeTotalContainer.stakeChangePeriods.push(nextPeriod);
+    } else if (stakeTotalContainer.stakeChangePeriods.length == 0) {
+      stakeTotalContainer.stakeChangePeriods.push(nextPeriod);
+    }
+
+    if (stakeUserContainer.stakeChangePeriods.length != 0) {
+      uint16 lastStakePeriod = stakeUserContainer.stakeChangePeriods[stakeUserContainer.stakeChangePeriods.length - 1];
+      if (lastStakePeriod != nextPeriod) {
+        stakeUserContainer.stakeChangePeriods.push(nextPeriod);
+      }
+
+      if (stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount > stakeTokens) {
+        stakeTotal.totalStakedAmount += stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount - stakeTokens;
+        if (stakeTokens == 0) stakeTotal.stakingUsersCount--;
+        stakeUserContainer.userStake[nextPeriod].changedStake = stakeUserContainer.userStake[lastStakePeriod].totalStakedAmount - stakeTokens;
```

Comment:
What if user changes stake for period multiple times? Will `changedStake` be correct?